### PR TITLE
fix(js_parser): handle edge cases for arrow expression in tenary

### PR DESCRIPTION
**Pull request by kalleep**
## Summary
When looking at what caused the issues describe in #1077 I first noticed that the consequent block was parsed as a arrow expression because when we try to parse an arrow function we also check for typescript return parameter. 

E.g for  `a ? (b) : <p>{() => null}</p>;`  everything after `:` and up until `=>` was parsed as the return parameter resulting in [this](https://github.com/biomejs/biome/blob/main/crates/biome_js_parser/src/syntax/function.rs#L599) being true.

My first try to fix this was to return an error when `ambiguity.is_disallowed() ` was true and if the parsed typescript return parameter was bogus.

While this worked for this perticular case I started to play around with different tenary expressions that could generate the same result but with valid typescript syntax and found:
```javacsript
a ? (b) : a => {};
```

This expression failed both when parsing typescript and javascript but with different errors and checking for bogus typescript return type was not fixing this case.

So instead I created a specialized version of  `parse_assignment_expression_or_higher` to use when parsing the consequent block of a tenary expression.

Playground link:
https://biomejs.dev/playground/?indentStyle=space&code=YQAgAD8AIAAoAGIAKQAgADoAIABhACAAPQA%2BACAAewB9ADsACgBhACAAPwAgACgAYgApACAAOgAgADwAcAA%2BAHsAKAApACAAPQA%2BACAAbgB1AGwAbAB9ADwALwBwAD4AOwA%3D

Closes: https://github.com/biomejs/biome/issues/1077

## Test Plan
Added new test

### CODE REVIEW

File: crates/biome_js_parser/src/syntax/expr.rs [hunk 1]
```diff
@@ -433,6 +434,32 @@ pub(super) fn parse_conditional_expr(p: &mut JsParser, context: ExpressionContex
...
+/// the next token is `:`.
+// test js arrow_expr_in_alternative
+// a ? (b) : a => {};
+
+// test ts ts_arrow_exrp_in_alternative
+// a ? (b) : a => {};
+
+// test jsx jsx_arrow_exrp_in_alternative
+// bar ? (foo) : (<a>{() => {}}</a>);
+fn parse_conditional_expr_consequence(
```
Comment (Conaclos) [inline crates/biome_js_parser/src/syntax/expr.rs:448]: consequent?
```suggestion
fn parse_conditional_expr_consequent(
```

File: crates/biome_js_parser/src/syntax/expr.rs [hunk 2]
```diff
@@ -433,6 +434,32 @@ pub(super) fn parse_conditional_expr(p: &mut JsParser, context: ExpressionContex
     }
 }
 
+/// Specialized version of [parse_assignment_expression_or_higher].
+/// We need to make sure that on a successful arrow_expression parse that
+/// the next token is `:`.
+// test js arrow_expr_in_alternative
```
Comment (Conaclos) [inline crates/biome_js_parser/src/syntax/expr.rs:440]: alternate?
```suggestion
// test js arrow_expr_in_alternate
```

**Review by Conaclos — APPROVED** [review]
This looks good to me! Thanks for your contribution!

I just suggested renaming `consequence` and `alternative` to `consequent` and `alternate`. Are there some specific reasons you chose `consequence` and `alternative`?

**Comment by kalleep** [comment]
No just mixing up some words

**Review by ematipico — APPROVED** [review]

**Pull request merged**